Implementations of Copy file to container command

vuminhkh opened the pull request:
Hello,

I've implemented "Copy file to container command", but it seems that migrating to 1.20 API breaks other part (8 tests failure). The commands that I modified, the test works well.

As I don't have enough knowledge of the project to fix them all, I still perform the pull request, hoping that it will be useful.

Thanks,

Minh Khang VU

marcuslinke commented:
@vuminhkh Thanks for your contribution. I will merge ASAP.

KostyaSha commented on src/main/java/com/github/dockerjava/core/CompressArchiveUtil.java:
```diff
@@ -2,18 +2,77 @@
+    static void putTarEntry(TarArchiveOutputStream tarOutputStream, TarArchiveEntry tarEntry, Path file) throws IOException {
+        tarEntry.setSize(Files.size(file));
+        tarOutputStream.putArchiveEntry(tarEntry);
+        InputStream input = new BufferedInputStream(Files.newInputStream(file));
```
Place in try-with-resources?

vuminhkh replied:
Thank you. It's better to go with new Java 7 functionality. It's cleaner.

KostyaSha commented on src/main/java/com/github/dockerjava/core/DockerClientImpl.java:
```diff
@@ -1,5 +1,12 @@
 package com.github.dockerjava.core;
 
+import static com.google.common.base.Preconditions.checkNotNull;
```
static usually imported after all imports

vuminhkh replied:
I'm using Intellij, it's re-arranged automatically like that by default. Please feel free to re-organize.

KostyaSha commented on src/main/java/com/github/dockerjava/core/DockerClientImpl.java:
```diff
@@ -77,16 +86,8 @@
 import com.github.dockerjava.core.command.VersionCmdImpl;
 import com.github.dockerjava.core.command.WaitContainerCmdImpl;
 
-import java.io.Closeable;
```
why do you reorder imports, what IDE do you use?

vuminhkh replied:
I'm using Intellij, it's re-arranged automatically like that by default. Please feel free to re-organize.

KostyaSha commented:
I see many unclosed streams, @vuminhkh could you check that you have no resource leaks https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html ?

vuminhkh commented:
Hello,

For streams unclosed, if i'm not mistaken it's about the tar method in CompressArchiveUtil, as it's stream which wrap stream, closing the most outer will delegate to inner and the whole stream wrapped chain will be closed.

For formatting rules, i'm using Intellij, please feel free to re-organize based on your IDE settings, or give me some hints to have the same format rules as yours.

Thanks,

Minh Khang VU

marcuslinke commented:
@vuminhkh After thinking about again we should stay backward compatible. In fact the `CopyFileFromContainerCmd` should target to `/containers/{id}/copy` and then
a new `CopyArchiveFromContainerCmd` as well as a `CopyArchiveToContainerCmd` should target to `
/containers/{id}/archive`. Would it be possible for you to change your PR accordingly please?

vuminhkh commented:
Hi Marcus,

I'll find my-self some times to do this asap.

Minh Khang VU

vuminhkh commented:
Hi Marcus,

Please review to see if it's ok. There are always 8 test failures (the same as before) but they don't concern the commands that I added/modified.

Regards,

Minh Khang VU
[TestSuite.txt](https://github.com/docker-java/docker-java/files/42201/TestSuite.txt)

KostyaSha commented on src/main/java/com/github/dockerjava/api/command/CopyArchiveFromContainerCmd.java:
```diff
@@ -0,0 +1,38 @@
+
+    public static interface Exec extends DockerCmdSyncExec<CopyArchiveFromContainerCmd, InputStream> {
+    }
+}
```
new line in the end of file

KostyaSha commented on src/main/java/com/github/dockerjava/api/DockerClient.java:
```diff
@@ -119,8 +121,12 @@
 
     public LogContainerCmd logContainerCmd(String containerId);
 
+    public CopyArchiveFromContainerCmd copyArchiveFromContainerCmd(String containerId, String resource);
```
please add javadoc and mention from what docker-api version it works.

KostyaSha commented on src/main/java/com/github/dockerjava/api/command/CopyArchiveToContainerCmd.java:
```diff
@@ -0,0 +1,60 @@
+     * Set container's id
+     *
+     * @param containerId id of the container to copy file to
+     * @return this
```
remove return description, it makes no sense

KostyaSha commented on src/main/java/com/github/dockerjava/core/command/CopyFileFromContainerCmdImpl.java:
```diff
@@ -74,4 +74,4 @@ public CopyFileFromContainerCmdImpl withHostPath(String hostPath) {
         return super.exec();
     }
-}
+}
```
unrelated to PR change

KostyaSha commented on src/main/java/com/github/dockerjava/core/util/FilePathUtil.java:
```diff
@@ -12,11 +13,9 @@ private FilePathUtil() {
-     * @param file
-     *            the file to get the relative path
+     *
+     * @param baseDir a parent directory of {@code file}
```
unrelated change

KostyaSha commented:
@vuminhkh please exclude all unrelated to your change formatting and import changes

vuminhkh commented:
@KostyaSha I tried to use your formatting rules at docker-java/etc/code-style.epf and eclipse plugin for intellij, it seems that i's not your current formatting rules. So I made all the changes "manually", what do you advice about formatting rules in order to contribute to your project ?

marcuslinke commented:
@vuminhkh Thanks! Hopefully I'm able to merge this evening...

marcuslinke commented:
@vuminhkh After looking deeper into this I've just found that the API of `CopyArchiveToContainerCmd` prevents from copying an already compressed tar archive to the container. This is useful if someone wants to copy an archive from one container to another one.

The command should be able to transport a `InputStream` to the `CopyArchiveToContainerCmdExec` just like the `BuildImageCmd` does. The tar compression must be done in the command and not in the exec class then. Would it be possible for you to adopt your PR accordingly please?

Beside this, the NullpointerExceptions that you got are caused by https://github.com/docker-java/docker-java/issues/388

KostyaSha commented on src/main/java/com/github/dockerjava/api/DockerClient.java:
```diff
@@ -119,8 +121,37 @@
+
+    /**
+     * Copy resource from container to local machine.
+     * Deprecated, since docker API version 1.20 and above use copyArchiveFromContainerCmd instead.
```
`@deprecated` javadoc

KostyaSha commented on src/main/java/com/github/dockerjava/jaxrs/DockerCmdExecFactoryImpl.java:
```diff
@@ -173,7 +177,7 @@ public void init(DockerClientConfig dockerClientConfig) {
 
     private org.apache.http.config.Registry<ConnectionSocketFactory> getSchemeRegistry(final URI originalUri,
-            SSLContext sslContext) {
+                                                                                       SSLContext sslContext) {
```
unrelated change

KostyaSha commented on src/test/java/com/github/dockerjava/core/command/CopyFileFromContainerCmdImplTest.java:
```diff
@@ -70,4 +70,4 @@ public void copyFromNonExistingContainer() throws Exception {
         }
     }
-}
+}
```
unrelated change

vuminhkh commented:
@marcuslinke I agree with you, It's not much modification but the API is more generic. I updated the API according to your suggestion.
@KostyaSha API doc and deprecated annotation updated to Java standard. For indentation issues, I insist that you commit your formatting rules and I'll use it to reformat files that I modified.

Thanks,

Minh Khang VU

KostyaSha commented:
@vuminhkh i slowly adding checkstyle that matches to what marcus did. The standard rule for projects that has not yet defined styles is not mix styles and follow existed.

KostyaSha commented on src/main/java/com/github/dockerjava/api/DockerClient.java:
```diff
@@ -121,8 +123,37 @@
+     * @param resource    path to container's resource
+     * @return created command
+     * @see #copyArchiveFromContainerCmd(String, String)
+     * @deprecated since docker API version 1.20, replaced by {@link #copyArchiveFromContainerCmd(String, String)}
```
(Not critical, but for simplification there are API versions for linking javadocs. Example https://github.com/docker-java/docker-java/blob/58d94079d2cd7ae7d072e5aad599777707e5b10a/src/main/java/com/github/dockerjava/api/command/InspectContainerResponse.java#L234 )

KostyaSha replied:
Try change link to version here also

vuminhkh commented:
@KostyaSha updated javadoc following your hint.
@marcuslinke and @KostyaSha I began to implement new network API, you can preview at https://github.com/vuminhkh/docker-java/tree/network . I'll perform another pull request once this one merged.

Thanks,

Minh Khang VU

KostyaSha commented:
@vuminhkh implement classes is pretty simple, problem is wrote tests and verify this changes.

vuminhkh commented:
@KostyaSha For commands that I implemented, if i'm not mistaken, I've written tests for them all. Normally what I wrote did not break any existing feature, tests that are broken were APIs which are not compatible anymore with newer Docker API. I have not fixed those tests yet, I'll do it once I have more time. If you guy could help me also on this it will be great as those I don't have knowledge about those existing command.

vuminhkh commented:
@marcuslinke @KostyaSha For breaking changes in docker API, what will be your strategy to be compatible with multiple docker api version ?

KostyaSha commented:
Depends on cases.

2015-12-05 21:45 GMT+03:00 vuminhkh <notifications@github.com>:

> @marcuslinke <https://github.com/marcuslinke> @KostyaSha
> <https://github.com/KostyaSha> For breaking changes in docker API, what
> will be your strategy to be compatible with multiple docker api version ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker-java/docker-java/pull/347#issuecomment-162234167>
> .
>

vuminhkh commented:
@KostyaSha What is the configuration of your test ? Which version of docker do you target to assure that there's not regression ? If i'm not mistaken it's here https://github.com/docker-java/docker-java/blob/master/src/test/java/com/github/dockerjava/client/AbstractDockerClientTest.java#L41 that you define which version of API the test target. Is it ok if I change it to 1.21 ?

KostyaSha commented:
I'm bit newbie with current tests design. @marcuslinke please suggest.

marcuslinke commented:
@vuminhkh, @KostyaSha  Currently we can only test against one specific version of the remote API. This is also the reason for the compatibility statement here: https://github.com/docker-java/docker-java#docker-java

As you already found this is hard-coded in the `AbstractDockerClientTest` class. I will push some changes related to issue https://github.com/docker-java/docker-java/issues/388 soon that will change the tested API version.

marcuslinke commented:
After merging https://github.com/docker-java/docker-java/pull/392 I will take a look at your latest changes @vuminhkh.

marcuslinke commented:
@vuminhkh Thanks for contributing!
